### PR TITLE
fix(utils): Make nested dict uniqueness checks work

### PR DESCRIPTION
**Pull request by kattrali**
## Goal

Fix potentially over-aggressive recursion checking where references are discarded between id() checks.

## Design

References to nested dicts/field are recycled when going out of scope, which is immediately since the references are not retained for the lifetime of the sanitization pass. This change keeps references to seen nested objects in addition to ignored objects to ensure id() returns unique references.

Related reports on the Python issue tracker which explain the issue:
* https://bugs.python.org/issue36156
* https://bugs.python.org/issue33685

```
>>> help(id)

Help on built-in function id in module __builtin__:

id(...)
    id(object) -> integer

    Return the identity of an object.  This is guaranteed to be unique
    among simultaneously existing objects.  (Hint: it's the object's 
    memory address.)
```

## Changeset

* Update `_sanitize` and `_sanitize_dict` encoder methods to keep seen references

## Tests

* Added a new test case which checks for recursion in a large enough dict
* Tested against `tox` config and underlying `nosetests` directly

## Linked issues

Related to #179 
Closes #180

## Review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases

### CODE REVIEW

File: bugsnag/utils.py [hunk 1]
```diff
@@ -123,7 +128,7 @@ def _sanitize_dict_key_value(self, clean_dict, key, clean_value):
                     'Could not add sanitize key for dictionary, '
                     'dropping value.')
 
-    def _sanitize_dict(self, obj, trim_strings, ignored):
+    def _sanitize_dict(self, obj, trim_strings, ignored, seen):
```
[the remark this answers is not in the document]
Comment (kattrali) [reply bugsnag/utils.py]: yeah its intentional, since filtering does its own "ignored" checking as its available at the top level. Though it should also do its own "seen" checking!

File: bugsnag/utils.py [hunk 2]
```diff
@@ -81,13 +81,15 @@ def default(self, obj):
...
 
-    def _sanitize(self, obj, trim_strings, ignored=None):
+    def _sanitize(self, obj, trim_strings, ignored=None, seen=None):
         """
         Replace recursive values and trim strings longer than
         MAX_STRING_LENGTH
         """
         if not ignored:
             ignored = set()
+        if not seen:
```
Comment (tobyhs) [inline bugsnag/utils.py]: It seems preferable to compare `seen` to `None` directly, as `not []` is true, so it would create another empty array unnecessarily. However, I can't see how `_sanitize` will be called with an empty `seen` with the current code, so it probably doesn't matter too much.

File: bugsnag/utils.py [hunk 3]
```diff
@@ -81,13 +81,15 @@ def default(self, obj):
...
+            seen = []
```
Comment (tobyhs) [inline bugsnag/utils.py]: Consider adding a comment to explain why this `seen` parameter is needed, as it probably isn't obvious to someone reading the code without looking at the pull request description or commit message.

**Comment by kattrali** [comment]
Thanks for the review, @tobyhs, those sound like nice additions. Updated the PR to match.
